dimod/decorators.py

```diff
@@ -497,3 +498,20 @@ def wrapper(obj, *args, **kwargs):
         return method(*args, **kwargs)
 
     return wrapper
+
+def unique_variable_labels(f):
+    """Decorator to assign unique labels to variables when no label is passed.
+
+    Designed to be applied to variable methods, :meth:`.dimod.Binary`,
+    :meth:`.dimod.Spin`, :meth:`.dimod.Integer`.
+
+    """
+    @wraps(f)
+    def conditional_unique_label(label = None, *args, **kwargs):
+        if label is None:
+            qm = f(label=uuid.uuid4(), *args, **kwargs)
```

Comment:
I am actually a bit surprised that Python lets you order it this way. Since for 
```
def f(label=None, *args, **kwargs): pass
```
`f(label=1, 3)` raises a 
```
SyntaxError: positional argument follows keyword argument
```
and `f(3, label=1)` raises a
```
TypeError: f() got multiple values for argument 'label'
```.

You could do 
```
def conditional_unique_label(*args, label = None, **kwargs):
    ...
```
or (my preference) just drop `*args` altogether.

Reply:
On second thought, I suppose this is fine the way it is, since it allows us to add more keyword arguments in the future while maintaining the position of `label`.

Reply:
Those two errors are expected given the arguments can be passed either positionally, positionally followed by keyword (for further down the line positional args), or all as keyword args. We could consider making all of the args keyword only with defaults, but usage would be more verbose. And I don't think I can drop `*args`, because I still want the option to pass the positonal arguments that follow `label` i.e. `Spin('a',1)` should remain valid.

Reply:
```
>>> dimod.Binary('a',1) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: conditional_unique_label() takes from 0 to 1 positional arguments but 2 were given
```
This is the issue we would see without `*args`

Reply:
Ah, that is what I was missing. Ok, I stand corrected, thanks!